### PR TITLE
poetry2nix: init at 1.1.0

### PR DESCRIPTION
**Pull request by adisbladis**
###### Motivation for this change
This adds [poetry2nix](https://github.com/nix-community/poetry2nix) to nixpkgs and uses it to re-package `poetry`.

This new poetry derivation has a number of advantages over the old one:
- "Vendored" dependencies
Prevents clashes between nixpkgs `poetry` and projects dependencies
- Uses `poetry`'s own lock files
This ensures a well-tested set of dependencies.
- Should be _much_ easier to keep up-to-date

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notify maintainers

cc @zimbatm @gilligan @andir

### CODE REVIEW

**Comment by adisbladis** [comment]
@GrahamcOfBorg build poetry

File: pkgs/development/tools/poetry2nix/poetry2nix/README.md [hunk 1]
```diff
@@ -0,0 +1,2 @@
+Dont change these files here, they are maintained at:
```
Comment (FRidh) [inline pkgs/development/tools/poetry2nix/poetry2nix/README.md]: While development is done there, do describe the process for updating these expressions
pkgs/development/tools/poetry2nix/update

File: pkgs/development/python-modules/poetry/default.nix [hunk 2]
```diff
@@ -1,93 +1,64 @@
...
-    html5lib
-    shellingham
-    tomlkit
-  ] ++ lib.optionals (isPy27 || isPy34) [ typing pathlib2 glob2 ]
-    ++ lib.optionals isPy27 [ virtualenv functools32 subprocess32 ];
+    mkdir -p $out/bin
+    cat > $out/bin/poetry <<EOF
+    #!${runtimeShell}
+    export PYTHONPATH=$out/lib/${python.libPrefix}/site-packages:$PYTHONPATH
+    exec ${python}/bin/python -m poetry "\$@"
```
Comment (FRidh) [inline pkgs/development/python-modules/poetry/default.nix]: ${python.interpreter}

**Review by FRidh — CHANGES_REQUESTED** [review]

**Comment by FRidh** [comment]
While I think it is great for developers to be able to use Poetry in Nix for creating reproducible environments and applications (I will be using it myself), I am a bit skeptical towards using it in Nixpkgs.

Lock files make it quick and easy to package exactly what upstream intended, however, that's not always good. For example, they pin patch versions which means distributing security updates becomes harder. That, along with reporting to upstream issues we encounter is I think an important role of distributions. Clearly, this remark isn't just applicable to poetry2nix but also other such converters as well as Docker images.

**Comment by adisbladis** [comment]
> Lock files make it quick and easy to package exactly what upstream intended, however, that's not always good. For example, they pin patch versions which means distributing security updates becomes harder. 

While this is true there is nothing that says we need to use uptreams lock files, throwing away upstream's lock-file and running `poetry lock` is imho perfectly fine.
There is also the question of correctness, it's currently _really_ hard to ensure that bumping a python package does not cause breakage elsewhere.

Another thing about this: There are _a lot_ of applications out there with complex dependency graphs that I will never package in nixpkgs without `poetry2nix`.
I'm simply not interested in spending time on packaging [pretix](https://github.com/pretix/pretix/blob/1d36ef3/src/setup.py#L90) or [pretalx](https://github.com/pretalx/pretalx/blob/86b3ad0/src/setup.py#L93) without proper automation.

**Comment by zimbatm** [comment]
We could also use poetry2nix to maintain the current python packages set.

The more we can move towards pure data in nixpkgs, the more it will allow us to write higher-level tooling to automate all the grunt work. In the future, GitHub will be mostly used by bots talking to each-other while we sip margaritas at the beach (or do other more interesting things than manual dependency resolution).

**Comment by FRidh** [comment]
> While this is true there is nothing that says we need to use uptreams lock files, throwing away upstream's lock-file and running poetry lock is imho perfectly fine.

This is a good idea I think. So long-term we need a way to regenerate (all) lock files and make them use the latest patch releases.

>  There are a lot of applications out there with complex dependency graphs that I will never package in nixpkgs without poetry2nix.

Clearly. For these kinds of applications it just does not make sense to use only the main packages set.

> We could also use poetry2nix to maintain the current python packages set.

We could start a thread about this on Discourse but I can tell you already now it will fail with version conflicts when resolving dependencies (locking). Like now a way is needed to handle such situation.

**Comment by adisbladis** [comment]
I consider this merge-ready.

**Review by FRidh — APPROVED** [review]

**Comment by adisbladis** [comment]
@GrahamcOfBorg build poetry

**Review by jonringer — APPROVED** [review]

**Comment by knedlsepp** [comment]
This is great! Makes me want to introduce poetry in all my python projects. Is there any chance however that this did break `python36Packages.pyopenssl`?

**Comment by adisbladis** [comment]
@knedlsepp Nope, nothing in `pythonPackages` was touched in this PR.

**Comment by knedlsepp** [comment]
@adisbladis: Oh. Sorry about that, I only noticed that it did break somewhere between `62ef77...ec29bb` and this commit at least has `pyopenssl` in the `overrides.nix`. Verified that this is unrelated.

**Comment by asymmetric** [comment]
One of our [macOS Travis jobs](https://travis-ci.com/dapphub/dapptools/jobs/272345595#L579-L582) seems to be broken since this got merged.

The error is:

```
$ nix-env -i ShellCheck
grep -RHL --exclude-dir node_modules node src/*/libexec | xargs shellcheck --exclude=2001

error: attribute 'fromTOML' missing, at /nix/store/6vni58c49p2pwmpzkjr7zsf1b2arsd98-nixpkgs-20.03pre207234.ec29bb50bf4/nixpkgs/pkgs/development/tools/poetry2nix/poetry2nix/lib.nix:27:20

(use '--show-trace' to show detailed location information)
```

**Comment by adisbladis** [comment]
@asymmetric I suggest upgrading your nix version to 2.1 or newer.

**Comment by zimbatm** [comment]
Why is poetry2nix being evaluated at all? That sounds bad for performance..

**Comment by zimbatm** [comment]
here is a back-compat for fromTOML that uses IFD: https://github.com/nmattia/naersk/blob/7f30aa64ce25959b0872012d990060016ae64f88/builtins/default.nix#L12-L34

**Pull request merged**